### PR TITLE
[IMP] point_of_sale,pos_self_order: handling product attributes

### PR DESCRIPTION
**Pull request by vlst-odoo**
When selecting an attribute value in POS, what we store is the `description`,
which is a string that represents the selected attribute values.
Ex: selecting `Size: M` and `Material: Leather` will result in the description:
`(L, Leather)`.

This choice does not lead to a logical API for dealing with product attributes.

In this pr, we add a new field that stores the selected `ids` of
`"product.template.attribute.value"` and remove the
`selected_attributes` `Json` field from the `pos_self_order`
override of the `pos.order.line` model.

The `attributeHelper` function from the `tour_utils.js` file
from `pos_self_order` is improved such that it can now handle
both checking if a certain attribute is selected and actually
selecting an attribute.

Task: 3378533
https://github.com/odoo/upgrade/pull/4992

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/pos_self_order/models/pos_order.py [hunk 1]
```diff
@@ -10,17 +10,21 @@
...
-    # For the moment we need this to keep attributes consistency between the server and client_side.
-    selected_attributes = fields.Json(string="Selected Attributes")
-
     # FIXME: uuid already pass in pos and move note in pos_restaurant.
     def _export_for_ui(self, orderline):
         return {
-            'uuid': orderline.uuid,
-            'note': orderline.note,
+            "uuid": orderline.uuid,
+            "note": orderline.note,
```
Comment (adgu-odoo) [inline addons/pos_self_order/models/pos_order.py:17]: useless diff

Comment (davidmonnom) [reply addons/pos_self_order/models/pos_order.py:17]: @vlst-odoo In python, variable names must be surrounded by ' and in javascript by ". 
In python only the text intended for the user must be surrounded by ".

File: addons/point_of_sale/static/src/app/store/product_configurator_popup/product_configurator_popup.js [hunk 2]
```diff
@@ -74,25 +75,29 @@ export class ProductConfiguratorPopup extends AbstractAwaitablePopup {
 
     getPayload() {
         var selected_attributes = [];
+        var selected_attributes_ids = [];
```
[the remark this answers is not in the document]
Comment (vlst-odoo) [reply addons/point_of_sale/static/src/app/store/product_configurator_popup/product_configurator_popup.js:78]: i agree

File: addons/pos_self_order/models/pos_order.py (hunk 1 above)
Comment (vlst-odoo) [reply addons/pos_self_order/models/pos_order.py:17]: i didn't intentionally change it. That's just what the formatter does. I'll change it back anyways

File: addons/pos_self_order/models/pos_order.py [hunk 3]
```diff
@@ -20,6 +17,13 @@ def _export_for_ui(self, orderline):
             **super()._export_for_ui(orderline),
         }
 
+    def _get_selected_attributes(self) -> Dict:
+        """
+        :return: a dictionary that maps the attribute name to the attribute value name
```
Comment (adgu-odoo) [inline addons/pos_self_order/models/pos_order.py:22]: isn't it the ids and not name in the description ?

**Review by adgu-odoo — CHANGES_REQUESTED** [review]

**Comment by caburj** [comment]
- an order with desk organizer
- select attributes
- refresh the app
- pay the order
- checked the db but the attribute_value_ids is not in the order

Screenshot 2023-07-06 at 14 06 20

**Comment by caburj** [comment]
We can proceed. Kindly squash the commits.

**Comment by vlst-odoo** [comment]
> We can proceed. Kindly squash the commits.

done :+1:

**Review by caburj — APPROVED** [review]
robodoo r+

**Comment by caburj** [comment]
@vlst-odoo Kindly rebase and we'll retry.

**Comment by vlst-odoo** [comment]
> @vlst-odoo Kindly rebase and we'll retry.

It's done!

**Comment by vlst-odoo** [comment]
> @vlst-odoo Kindly rebase and we'll retry.

is it ok now?

**Comment by caburj** [comment]
robodoo r+

**Comment by caburj** [comment]
@vlst-odoo Can you check this? It could be a random error. We need to make sure it's not.

**Comment by vlst-odoo** [comment]
> @vlst-odoo Can you check this? It could be a ran

I don't get any error by running the test locally.

File: addons/pos_self_order/static/tests/tours/tour_utils.js [hunk 4]
```diff
@@ -135,86 +150,101 @@ export const PosSelf = {
...
+ * @returns {TourStep[]}
+ */
+export function attributesHelper(attributes, { isCheck = false } = {}) {
+    if (!attributes) {
+        return [];
     }
-
-    return attributesSteps;
-};
+    return attributes.map((attribute) => attributeHelper(attribute, isCheck));
```
Comment (vlst-odoo) [inline addons/pos_self_order/static/tests/tours/tour_utils.js:223]: inline this and remove it's function

File: addons/pos_self_order/static/tests/tours/tour_utils.js [hunk 5]
```diff
@@ -135,86 +150,101 @@ export const PosSelf = {
...
-    return increaseQuantity;
+// This function is used to generate the steps to increase / decrease the quantity of a product.
+const quantityHelper = (quantity) => {
+    // If the quantity is greater than 0 we decrease the `numOfClicks` by 1
+    const numOfClicks = Math.abs(quantity - 1 * (quantity > 0));
+    return [...Array(numOfClicks).keys()].map(() => ({
+        content: `${quantity > 0 ? "Increase" : "Decrease"} quantity`,
+        trigger: `.o_self_order_incr_button .btn:contains('${quantity > 0 ? "+" : "-"}')`,
+    }));
 };
```
Comment (vlst-odoo) [inline addons/pos_self_order/static/tests/tours/tour_utils.js:243]: remove diff

**Comment by vlst-odoo** [comment]
> The last paragraph of the commit message is no longer updated because "attributesHelper" was removed. Also, can you change the commit title to just list the modules that are changed. Make sure to adapt the commit message -- remove `pos*: ...`.
> 
> ```
> [IMP] point_of_sale,pos_self_order: handling product attributes
> ```
> 
> I believe we can proceed with this PR after this requested adjustments in the commit message and title.

:+1: done

**Review by caburj — APPROVED** [review]
robodoo r+

**Comment by caburj** [comment]
robodoo r+

**Review by caburj — APPROVED** [review]
robodoo r+

**Comment by caburj** [comment]
robodoo r+ (please)

**Pull request closed without merging**